app/api/resources/mentorship_relation.py

```diff
@@ -18,7 +18,56 @@
 from app.api.email_utils import send_email_mentorship_relation_accepted
 from app.api.email_utils import send_email_new_request
 
-mentorship_relation_ns = Namespace(
+
+# tools for inheritance
+from functools import wraps
+from flask import current_app, has_app_context
+from flask_restplus.utils import unpack
+from flask_restplus.marshalling import  marshal_with
+from flask_restplus.utils import merge
+
+class sub_marshal_with(marshal_with):
```

Comment:
What is this class for? 🤔 @hongsungin92

Reply:
@mentorship_relation_ns.marshal_list_with(mentorship_request_response_body) in line 211

  

the decorate

Reply:
when I makes a change 
" return [], HTTPStatus.BAD_REQUEST  ---> return messages.RELATION_STATE_FILTER_IS_INVALID, HTTPStatus.BAD_REQUEST " 
( file app/api/dao/mentorship_relation.py  in line 155)

" @mentorship_relation_ns.marshal_list_with(mentorship_request_response_body) "
( file app/api/resources/mentorship_relation.py in line 211)
this decorater makes a function 'get' of a class 'GetAllMyMentorshipRelation' (in line 180) unable to output the messages.RELATION_STATE_FILTER_IS_INVALID
it makes the function 'get' returns mentorship_request_response_body as below
 {
  "id": null,
  "action_user_id": null,
  "sent_by_me": null,
  "mentor": {
    "id": null,
    "name": null
  },
  "mentee": {
    "id": null,
    "name": null
  },
  "creation_date": null,
  "accept_date": null,
  "start_date": null,
  "end_date": null,
  "state": null,
  "notes": null
}

so I had to make the decorator to be applied only when a relation_state is valid 

the decorator function 'marshal_list_with' of a class 'Namespace' in flask_restplus.namespace  is as follows.
def marshal_list_with(self, fields, **kwargs):	
        return self.marshal_with(fields, True, **kwargs)

and the returned function 'marshal_with' of the class 'Namespace' is as follows.
    def marshal_with(self, fields, as_list=False, code=HTTPStatus.OK, description=None, **kwargs):
                def wrapper(func):
            doc = {
                'responses': {
                    code: (description, [fields]) if as_list else (description, fields)
                },
                '__mask__': kwargs.get('mask', True),  
            }
            func.__apidoc__ = merge(getattr(func, '__apidoc__', {}), doc)
            return marshal_with(fields, ordered=self.ordered, **kwargs)(func)
        return wrapper

and the returned class 'marshal_with' in flask_restplus.marshalling is as follows.
class marshal_with(object):
    def __init__(self, fields, envelope=None, skip_none=False, mask=None, ordered=False):
        self.fields = fields
        self.envelope = envelope
        self.skip_none = skip_none
        self.ordered = ordered
        self.mask = Mask(mask, skip=True)
    def __call__(self, f):
        @wraps(f)
        def wrapper(*args, **kwargs):
            resp = f(*args, **kwargs)
            mask = self.mask
            if has_app_context():
                mask_header = current_app.config['RESTPLUS_MASK_HEADER']
                mask = request.headers.get(mask_header) or mask
            if isinstance(resp, tuple):
                data, code, headers = unpack(resp)
                return (
                    marshal(data, self.fields, self.envelope, self.skip_none, mask, self.ordered),
                    code,
                    headers
                )
            else:
                return marshal(resp, self.fields, self.envelope, self.skip_none, mask, self.ordered)
        return wrapper

I override the function '__call__' of class 'marshal_with' just by adding " if ~ else ... " in the function 
" 
if HTTPStatus.OK in f(*args, **kwargs): <proceed with the original operation> 
else:  return f(*args, **kwargs)    #just output the decorated function 'get' 's return 

"

so class 'sub_marshal_with' is class 'marshal_with' having the overridden function '__call__'

Reply:
when I makes a change 
`return [], HTTPStatus.BAD_REQUEST`  ---> 
`return messages.RELATION_STATE_FILTER_IS_INVALID, HTTPStatus.BAD_REQUEST ` 
( file app/api/dao/mentorship_relation.py  in line 155)

`@mentorship_relation_ns.marshal_list_with(mentorship_request_response_body) `
( file app/api/resources/mentorship_relation.py in line 211)
this decorater makes a function `get` of a class `GetAllMyMentorshipRelation` (in line 180) unable to output the `messages.RELATION_STATE_FILTER_IS_INVALID`
it makes the function `get` returns `mentorship_request_response_body` as below
 {
  "id": null,
  "action_user_id": null,
  "sent_by_me": null,
  "mentor": {
    "id": null,
    "name": null
  },
  "mentee": {
    "id": null,
    "name": null
  },
  "creation_date": null,
  "accept_date": null,
  "start_date": null,
  "end_date": null,
  "state": null,
  "notes": null
}

so I had to make the decorator to be applied only when a `relation_state` is valid 

the decorator function `marshal_list_with` of a class `Namespace` in flask_restplus.namespace  is as follows.
```
def marshal_list_with(self, fields, **kwargs):	
        return self.marshal_with(fields, True, **kwargs)
```
and the returned function 'marshal_with' of the class 'Namespace' is as follows.
```
def marshal_with(self, fields, as_list=False, code=HTTPStatus.OK, description=None, **kwargs):
    def wrapper(func):
        doc = {
            'responses': {
                code: (description, [fields]) if as_list else (description, fields)
            },
            '__mask__': kwargs.get('mask', True),  # Mask values can't be determined outside app context
        }
        func.__apidoc__ = merge(getattr(func, '__apidoc__', {}), doc)
        return marshal_with(fields, ordered=self.ordered, **kwargs)(func)
    return wrapper
```
and the returned class `marshal_with` in flask_restplus.marshalling is as follows.
```
class marshal_with(object):
    def __init__(self, fields, envelope=None, skip_none=False, mask=None, ordered=False):
        self.fields = fields
        self.envelope = envelope
        self.skip_none = skip_none
        self.ordered = ordered
        self.mask = Mask(mask, skip=True)
    def __call__(self, f):
        @wraps(f)
        def wrapper(*args, **kwargs):
            resp = f(*args, **kwargs)
            mask = self.mask
            if has_app_context():
                mask_header = current_app.config['RESTPLUS_MASK_HEADER']
                mask = request.headers.get(mask_header) or mask
            if isinstance(resp, tuple):
                data, code, headers = unpack(resp)
                return (
                    marshal(data, self.fields, self.envelope, self.skip_none, mask, self.ordered),
                    code,
                    headers
                )
            else:
                return marshal(resp, self.fields, self.envelope, self.skip_none, mask, self.ordered)
        return wrapper
```
I override the function `__call__` of class `marshal_with` just by adding `if ~ else ...` in the function 
```
if HTTPStatus.OK in f(*args, **kwargs): <proceed with the original operation> 
else:  return f(*args, **kwargs)    #just output the decorated function 'get' 's return 
```

so class `sub_marshal_with` is class `marshal_with` having the overridden function `__call__`

Reply:
I think the solution is a little messy. let me know if there is a problem with this solution!

Reply:
Thank you for providing such a deep explanation @hongsungin92 👏 I really appreciate that. I will ask for help because I am a bit overwhelmed with life/work.

Reply:
Okay @isabelcosta. Cheer up